mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td

```diff
@@ -445,7 +414,7 @@ def LLVM_StoreOp : LLVM_MemAccessOpBase<"store",
   let assemblyFormat = [{
     (`volatile` $volatile_^)? $value `,` $addr
     (`atomic` (`syncscope` `(` $syncscope^ `)`)? $ordering^)?
-    attr-dict `:` custom<StoreType>(type($value), type($addr))
+    attr-dict `:` type($value) `,` qualified(type($addr))
```

Comment:
Why do we need `qualified` here?

Reply:
Without `qualified`, this just prints `<1>`, instead of `!llvm.ptr<1>`. Tbh, I do not fully understand the logic behind this, but I've observed similar issues in other contexts.

Reply:
@ftynse are you fine with this change or should we get to the bottom of this issue?

Reply:
I'd like to understand better why and when is this necessary (the feature looks poorly documented), but this should not block this from landing.

Reply:
This behaviour was introduced by https://github.com/llvm/llvm-project/commit/ee0908703d2917d7310b71c5078fef44e8270317. As far as I can tell it essentially always omits the dialect prefix if the attribute/type class being parsed is known. Since in this Op it is known to be a `LLVMPointerType` it'll elide the `!llvm.ptr` by default when printing. 
https://github.com/llvm/llvm-project/commit/63f0c00d38ee7879239975a6743d4e6c7847b725 then apparently introduced the `qualified` directive to be able to restore the behaviour to before where the dialect prefix is always printed.

Personally I think this is non-intuative and that the default should be printing the full type by default instead, but that is another discussion entirely. Using `qualified`  seems to be the intended way to fully qualify the type in the assembly format.

Reply:
This is WAI: qualified is a poor default IMO.
For example why would `!llvm` be a better choice here? And even if it is, why shouldn't we express the choice?
Should we even print the type when the address space is the default one? Lot of possible simplifications...

Reply:
The problem is that not only `!llvm` is omitted, but the `ptr` prefix as well. It therefore prints just `<1>` when using an address space rather than either `ptr<1>` or `!llvm.ptr<1>`. This IMO makes it non-obvious what the type is, nor what the meaning of the `<1>` is supposed to be.

Reply:
Of course! But the answer is contextual...
For example it's not obvious to me that qualified is just the best answer here, what about:

```
let assemblyFormat = ` ....    attr-dict `:` type($value) `,` `ptr` `` type($addr)`
```

?

Reply:
The main reason I added `qualified` in this revision is to ensure as few breaking changes as possible, considering that the pointer type changes already forced me to change hundreds of lines.

I agree that `qualified` might not be the best idea here, as the "llvm." part brings no benefit. The format you propose would in my opinion be a much better fit. The question remains if we should always print the address space or not, and if we should even print the pointer type if the address space is zero.

Note that we should also consider all other operations where I used `qualified` in this discussion, i.e., `LLVM_StoreOp`, `LLVM_AtomicRMWOp`, `LLVM_AtomicCmpXchgOp`, `NVVM_MBarrierArriveSharedOp`.

Reply:
To be clear I wasn't criticizing this PR: I was providing a rational for "building new dialects".